Mathlib/CategoryTheory/Sites/Preserves.lean

```diff
@@ -0,0 +1,179 @@
+/-
+Copyright (c) 2023 Dagur Asgeirsson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Dagur Asgeirsson
+-/
+import Mathlib.CategoryTheory.Limits.Preserves.Shapes.Products
+import Mathlib.CategoryTheory.Limits.Shapes.CommSq
+import Mathlib.CategoryTheory.Sites.EqualizerSheafCondition
+
+/-!
+# Sheaves preserve products
+
+We prove that a presheaf which satisfies the sheaf condition with respect to certain presieves
+preserve "the corresponding products".
+
+## Main results
+
+More precisely, given a presheaf `F : Cᵒᵖ ⥤ Type*`, we have:
+
+* If `F` satisfies the sheaf condition with respect to the empty sieve on the initial object of `C`,
+  then `F` preserves terminal objects.
+See `preservesTerminalOfIsSheafForEmpty`.
+
+* If `F` furthermore satisfies the sheaf condition with respect to the presieve consisting of the
+  inclusion arrows in a coproduct in `C`, then `F` preserves the corresponding product.
+See `preservesProductOfIsSheafFor`.
+
+* If `F` preserves a product, then it satisfies the sheaf condition with respect to the
+  corresponding presieve of arrows.
+See `isSheafFor_of_preservesProduct`.
+-/
+
+universe v u
+
+namespace CategoryTheory.Presieve
+
+variable {C : Type u} [Category.{v} C] {I : C} (F : Cᵒᵖ ⥤ Type (max u v))
+
+open Limits Opposite
+
+variable (hF : (ofArrows (X := I) Empty.elim instIsEmptyEmpty.elim).IsSheafFor F)
+
+section Terminal
+
+variable (I) in
+/--
+If `F` is a presheaf which satisfies the sheaf condition with respect to the empty presieve on any
+object, then `F` takes that object to the terminal object.
+-/
+noncomputable
+def isTerminal_of_isSheafFor_empty_presieve : IsTerminal (F.obj (op I)) := by
+  refine @IsTerminal.ofUnique _ _ _ fun Y ↦ ?_
+  choose t h using hF (by tauto) (by tauto)
+  exact ⟨⟨fun _ ↦ t⟩, fun a ↦ by ext; exact h.2 _ (by tauto)⟩
+
+/--
+If `F` is a presheaf which satisfies the sheaf condition with respect to the empty presieve on the
+initial object, then `F` preserves terminal objects.
+-/
+noncomputable
+def preservesTerminalOfIsSheafForEmpty (hI : IsInitial I) : PreservesLimit (Functor.empty Cᵒᵖ) F :=
+  haveI := hI.hasInitial
+  (preservesTerminalOfIso F
+    ((F.mapIso (terminalIsoIsTerminal (terminalOpOfInitial initialIsInitial)) ≪≫
+    (F.mapIso (initialIsoIsInitial hI).symm.op) ≪≫
+    (terminalIsoIsTerminal (isTerminal_of_isSheafFor_empty_presieve I F hF)).symm)))
+
+end Terminal
+
+section Product
+
+variable (hI : IsInitial I)
+
+-- This is the data of a particular disjoint coproduct in `C`.
+variable {α : Type} (X : α → C) [HasCoproduct X] [(ofArrows X (Sigma.ι X)).hasPullbacks]
+    (hd : ∀ i j [HasInitial C], i ≠ j →
+      IsPullback (initial.to _) (initial.to _) (Sigma.ι X i) (Sigma.ι X j))
+    [∀ i, Mono (Sigma.ι X i)]
```

Comment:
I think that a more mathlib phrasing of the definitions and lemmas in this section would be to take as input `c : Cofan X` that is colimit. The `HasInitial` should be put out of the `hd` assumption. Then, `Sigma.ι` would be replaced by `c.inj`.

Reply:
I'll make this change later tonight or tomorrow.

Reply:
I started doing this, and `firstMap_eq_secondMap` is fine with the general cocone, but in the rest of the file I'm relying heavily on the particular choice of product/coproduct by using `piComparison`. So introducing this general cocone would mean that the approach would have to change a lot for the rest of the file.

Reply:
I am afraid this is the right approach here :-(

Reply:
I do not mind if the internals of the proof uses the `piComparison`, but the main lemmas like `isSheafFor_of_preservesProduct` and ` isSheafFor_iff_preservesProduct` should be more general. `isSheafFor_of_preservesProduct` is already quite close to what I would like to get: the data `π : (i : α) → X i ⟶ Z` is mostly that of a `Cofan`, and the assumption `[IsIso (Sigma.desc π)]` is equivalent to saying that the cofan is colimit. Then, for this lemma, in the worst case scenario, if we change the assumptions as I suggest, the proof *could* start by the proof that `IsIso (Sigma.desc π)` holds, and the rest of the proof should not have to change much.